### DISCUSSION TITLE
Custom domain and buying domains

### DISCUSSION THREAD

**Discussion by Danish2023github** [community/community#60334] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

Hello everyone! I am new to github and website development. I had published a basic structured website last night and it did work.
I got a default web address/domain from github. I then tried to make my own custom domain. I randomly made one. And it did not work.

1. Do I have to buy a domain from someone or can I randomly make a website address?
2. What is the complete method of making a custom domain without errors on github?
3. What are subdomains?

**Comment by Danish2023github** [comment]
![Annotation 2023-07-07 190251](https://github.com/community/community/assets/133471108/3981f481-b119-4aaa-9e50-1ecee30fdcc0)
This is what I got in notifications.

**Comment by Danish2023github** [comment]
![Annotation 2023-07-07 190251](https://github.com/community/community/assets/133471108/c8b737d4-3b03-49a7-a255-cfa8b02347a5)
This one too. Please assist me with this.

**Comment by Danish2023github** [comment]
I would like to mention that I did try to make different domains names. But when it would not work then i would simply remove them from pages option and it would get back to the default github provided domain.

**Comment by ronakjkothari** [comment, marked as answer]
I will address your questions in order:

1. Do I have to buy a domain from someone or can I randomly make a website address?
You must purchase a domain from certain vendors such as [NameCheap](https://www.namecheap.com/) or [GoDaddy](https://www.godaddy.com/). For test projects or unofficial websites, I would use the URL that GitHub pages created, however, if it is for an official website (such as my website [Elephant - The Ultimate Student Suite](https://elephantsuite.net/)) I would consider investing in a domain. Some of the domains, depending on their rarity, are expensive, so be warned.

2. What is the complete method of making a custom domain without errors on GitHub?
Follow the guide provided by [this YouTube video](https://www.youtube.com/watch?v=2K7asqt8wMw)

3. What are subdomains?
They are like those things preceding the actual domain. Kind of like [https://docs.elephantsuite.net/](https://docs.elephantsuite.net/). Elephantsuite.net is the actual domain, but there is the "docs" subdomain.


Hope this helps!

**Reply by Danish2023github** [reply]
Thank you so much bro. I really appreciate your detailed answer. Now you have answered my domain related questions, so I have other questions related to the purchase of domain.

1. Once I purchase a domain, does it become my property? In other words, do I have to pay yearly or monthly premium for a domain?
2. Second, is domain a permanent name or can I change it as per my requirement? For instance, a person buys google.com as a domain name. Can he change that domain name to something like apple.com? Or it has to remain Google.com at all cost?

**Reply by ronakjkothari** [reply]
No problem for the answer! As for the payment, there is a monthly/yearly payment for most if not all domains. Domain names are permanent, however, if you wanted to rebrand from google.com to apple.com, you would want to sell the google.com domain (which would sell for a fortune), and purchase the apple.com domain (which would cost a fortune) and then move the HTML files to associate with apple.com. It's sort of like purchasing a real asset that can appreciate and depreciate over time.

**Reply by Danish2023github** [reply]
Wow! 

So, how do these domain providers or sellers like godaddy and others get all these domains? Do they make them? If they do, then how? I mean, do they not have to buy these domains they sell to us? 

Is domain like a fruit on a tree? Godaddy has the tree and they can get fruit(domains) from that tree?

**Reply by Danish2023github** [reply]
Oh and before I forget I must ask you about the subdomain function on the domains that we purchase from domain sellers. So can we also make subdomains of a domain that we purchase?

I was also curious about making subdomains of the github provided free or default domain. Can we make make subdomains of it?

**Reply by ronakjkothari** [reply]
Domain sellers like GoDaddy and NameCheap are pretty much like brokers. They don't own the domains, but they will do the heavy lifting of managing and completing the transaction of a domain. They make purchasing a domain easy and it makes managing a domain even easier.

As for subdomains, you can make subdomains of a domain that you purchase. [Here is an article](https://www.namecheap.com/support/knowledgebase/article.aspx/9776/2237/how-to-create-a-subdomain-for-my-domain/) that shows how to do it on NameCheap. Surely something similar can be done on GoDaddy. I myself prefer NameCheap.

I don't think you can make a subdomain from the free/default domain that is given.

**Reply by Danish2023github** [reply]
Thanks a lot bro! 
A. As I am new to coding and currently learning html and css on VS Code application on windows.
1. Why do I need to learn these codes when I can get templates of any kind of website with its own designs? 
2. While learning these codes, I kind of feel a little bit lost. How am I going to develop websites that look very beautiful and elegant as online shopping websites or websites of some mobile phone company?
3. Since there are many functions and their attributes, both in html and css and I am pretty sure there is a lot ahead in java and others as well, how do I remember all these codes? Do I really have to commit all these codes to my memory?

**Reply by ronakjkothari** [reply]
Glad you're beginning programming in HTML, CSS, and JavaScript. I have been programming websites for quite some time now. Learning HTML, CSS, and JavaScript is good for customization. There is only so much customization you can do with a template that you can barely customize. While they do most of the heavy lifting, you can barely customize the functionality. For example, making an ability to log in to your website using something like [SquareSpace](https://www.squarespace.com/), [Wix](https://www.wix.com/), or [Weebly](https://www.weebly.com/) would be sort of impossible. That's why learning these languages are good. Plus learning programming in this day and age is also great as JavaScript is becoming a mainstream programming language similar to Python.

Not shaming those websites, as they fill many needs. However, there are upsides to learning development yourself.

It is easy to get lost in website development. I have trouble learning a singular programming language, but in website development, you are learning 3 different languages. Focusing on beauty, I would recommend learning layouts (display:flex; or display:grid) in HTML and CSS, as that will help you with developing beautiful websites.

Not sure how new you are to development, however, I like to think of development as similar to reconstructing a human being. HTML is the bones, as it gives the website structure and is the foundation of the entire website. CSS is the skin and features, as it is responsible for making the website look fantastic and unique. JavaScript (JS) is like the organs, as it is responsible for giving life to the website and making it responsive. Putting them all together and you get a properly functioning website.

With committing them to memory, I would be honest, you probably have to commit to memory. It is more fun if you work on a project and try to make it looks good (look at layouts). I would say there are only a few CSS properties that you should be learning, as these are what I use all the time

```CSS
- display: flex;
- display: grid; (I don't use it often, however, I probably should)
- background-color: color;
- border-radius: pixelAmount;
- border: pixelAmount type (solid/dashed/dotted) color;
- box-shadow: pixelAmount pixelAmount pixelAmount;
- position: absolute;
- font-size: pixelAmount;
- font-family: "Arial";
- font-weight: Number between 100-900;
```

With HTML, `<div>` tags will be your best friend! Learning how to use Divs properly will boost your website design skills tenfold. Also learning the different header types and incorporating paragraphs and even buttons, inputs, and images will be good. I don't want to overwhelm but after mastering all of these (preferably with a project like designing an e-commerce website) you should be pretty good at Website Design. Then you can start tackling JavaScript

Hope this helps!

**Reply by Danish2023github** [reply]
It sure did help me bro. 
I will get back to you with more questions as I come across something new. 
And I am following you now. 
I will also keep your advice related to the above list of css properties and will surely practice them and commit them to memory.

**Comment by Danish2023github** [comment]
Bro, I have new questions today.
What is the difference between:
1. A relative position, and
2. Absolute position in css.

**Comment by ronakjkothari** [comment]
That's actually a good question. They do pretty similar things. To begin I need to explain the concept of a parent element (if this is familiar to you, feel free to skip)

A parent element is an element that is... the parent of another element. Pretty much an element that encompasses another element (see the example below)

```HTML
<div id="parent-element">
  <div id="child-element">
  </div>
</div>
```

That knowledge will be handy when working with `position:absolute;`. Anyway, `position:relative;` moves an element relative to itself. Pretty much you control the element's location based on where it is currently.

With `position:absolute;`, however, the element relies on the position specified for any parent element. In a scenario like the one given up top, if the parent-element was given position:absolute, then its position would be specified by the entire body of the website  (so a property like `top:0;` would bring the element to the top of the website) because that element has no parent with the position property specified. However, if both the parent-element and the child-element were given position:absolute, the child element would position itself based on the parent element's position, rather than the website entirely. This is true even if the parent had position:relative, rather than position:absolute;

This is useful to get layouts to behave how you want. See the example below where I used a combination of position:relative and position:absolute to achieve a pretty convincing badge modal.


![image](https://github.com/community/community/assets/70562508/de025d07-607e-4dd7-ac7f-ccbf25cf4e08)


If this is a little confusing, don't worry... it is a confusing topic. Hopefully [this article](https://kolosek.com/css-position-relative-vs-position-absolute) could clear it up better than I could.

**Reply by Danish2023github** [reply]
I think I got the point. 
1. Does this mean that absolute position gives an object absolute or total freedom to move around the webpage without having restrictions from margins or from being block item?

2. So if I have a parent div, which I had learnt yesterday and I am glad that I already had a clue about it, an item within that div with an absolute position would move within that div? Am I correct that the item would not be able to leave the div boundary no matter what number of top, bottom, left, or right I assign to it?

**Comment by ronakjkothari** [comment]
I'll answer the questions in order
1. For the most part yes 😄. It primarily depends on whether the parent div has a position:relative or position:absolute style attached to it.
2. Not necessarily. Firstly, if the parent div doesn't have a position:relative or a position:absolute, then the child div would actually move according to the website, therefore being able to go out of the div. Even if that parent div did have a position:relative or a position:absolute, you can actually attach negative numbers (for example top:-60px;) which will bring the child div outside the parent div.

**Reply by Danish2023github** [reply]
I get it. :)

**Reply by Danish2023github** [reply]
My next question is about the fonts of the website text.
I have learnt how to assign one or multiple fonts for a website text by making a link in the html and making an entry of the specific font name in the css file. 

But I need to know how about assigning fonts directly from the local storage on my pc. I am using vs code for learning these codings. I open a file from my local drive address and it opens up in the chrome browser. I can connect pictures to my website by putting a refernce of the storage address from within the pc. 

I downloaded a .ttf font file and I tried to add a link of the local storage in html but it didnt work.

can you guide me through this?

**Reply by ronakjkothari** [reply]
I've personally never liked downloading TTF fonts, as they are somewhat difficult to set up in comparison to using Google Fonts, however, I can guide you through it (it isn't that difficult, it's just a lot of room for error).

Place the TTF file in your workspace/directory/repository and then in your CSS file, link it using something similar to the below code

```CSS
@font-face {
    font-family: 'FontName';
    src: url('fontLinkInDirectory.ttf') format('truetype');
}  

body * {
    font-family: 'FontName', sans-serif;
}
```

Hopefully, this should work. BTW, the asterisk following the word `body` indicates that every single element in your HTML page regardless of parent will have the indicated `font-family` style added.

**Reply by Danish2023github** [reply]
Can I link the TTF file within the local storage? :)

**Reply by Danish2023github** [reply]
for instance, this is my url or link to my ttf file

"C:\Users\pc\Pictures\Sacramento\Sacramento-Regular.ttf"

**Reply by Danish2023github** [reply]
I have been linking all the stuff to my html by adding these links in vs code. when i add them to github, i replace the links by github repository links.

**Reply by Danish2023github** [reply]
Why I am doing this is because I dont want to depend on google fonts. I want to put all important things in my repository. even if ihave a need for gifs or anything, i would prefer to upload it to repository and link it to my website. I dont think it is reliable to link these images from other websites. 
I am afraid of these websites' unavailability.

**Reply by ronakjkothari** [reply]
I would say rather than putting the file in your Pictures folder, put it in the same folder as where your repository is on your computer. Then when you want to link it use something like `"./font.ttf"`

**Reply by Danish2023github** [reply]
I see. I will try it , too. I have to wrap this work up now and will continue tomorrow. I have spent much time on this. I will do it the way you have explained and then I would share with you the results. :)

**Reply by Danish2023github** [reply]
My bro, I was unable to try out the font property. I was away from the pc. I will try doing it asap and then if it doesnt work i will reach out to you. if it works i will inform you that it did work.

**Reply by ronakjkothari** [reply]
Sounds good. LMK where I can help 😄

**Comment by ronakjkothari** [comment]
I got an email pertaining to Google Fonts, but for some reason I am unable to see it. I can help guide you though.

Link your CSS to your HTML file. It would look something like this in your HEAD tag:

```HTML
<link rel="stylesheet" type="text/css" href="styles.css" />
//Replace styles.css with your css file path
```

Then in your CSS, copy paste this code:

```CSS
@import url('https://fonts.googleapis.com/css2?family=Montserrat:ital,wght@0,100;0,200;0,300;0,400;0,500;0,600;0,700;0,800;0,900;1,100;1,200;1,300;1,400;1,500;1,600;1,700;1,800;1,900&display=swap');

h1{
    font-family: 'Montserrat', sans-serif;
}
```

This should apply the Montserrat font to all h1 tags.

I'm glad you're into learning website development!

**Reply by Danish2023github** [reply]
Thank you so much, my dear brother. I have some more things to know about the websites:

1. After publishing my website on github pages, is it possible to use these websites for earnings by advertising? In other words, can we earn from the free github version?
2. How effective and useful web designing softwares, such as webflow, nicepage, figma are for a developer?

**Reply by ronakjkothari** [reply]
To my understanding, you should be able to monetize free github pages. I've never gone into that, but using Google AdSense should help with that.

There are two reasons I would see using web design software rather than conventional hard coding
1. You are a professional that has a lot of experience in UI/UX design, and Figma might help speed up the process by allowing designers to plan ahead.
2. You are planning a big project, and Figma allows you to get it right the first time.

If you are working on a small project, I would not recommend using Figma as design doesn't matter too much. But for much bigger projects, planning with Figma might be beneficial.

**Reply by Danish2023github** [reply]
I get it, my dear brother. I have a course on web development which has an inclination towards the traditional way, that is not to use figma stuff. 

But I recently got another course and this one stresses upon adopting new web development facilities, such as figma. 

But you have now explained very nicely what is suitable in what situation. 

Thank you so much, my dear brother once agin.

**Reply by Danish2023github** [reply]
All right. I will check the google adsense and see what I can get there.

**Comment by Danish2023github** [comment]
One more thing that is very important. I want to start working in the field of web development but i need some experience of the market.

How should i begin my professional journey as a web developer? 

I am not good at websites yet. I am still learning things. But I need some website orders by clients. 

Even if they are fake ones. I need fake website designs so that i can apply them for my experience.

Also, any ideas about what is a good approach to design websites? I mean we are learning development, but not the designs.

so how do i get these designs done. i have a course onthat too but have not checked it yet. what softwares do you prefer to use for designing web pages?

**Reply by ronakjkothari** [reply]
The way that I did it was by creating a copy of a preexisting website. If you feel you have somewhat mastered layouts (such as Flexbox and Grid), you can start tackling other websites. For example, recreate [Google](https://www.google.com/) and go on from there. Find a website that you like the design of, and recreate it. After you have done that a little and you understand how to use box shadow, transitions, and sizing properly, then try to create a website of your own.

Freelancing is a good way of beginning your professional journey. Either that or search Indeed for jobs like that. I would recommend putting together a portfolio of all your projects.

For design inspiration, I use [Dribbble](https://dribbble.com/). It is really good.

**Reply by Danish2023github** [reply]
Hello, Ronak. I was unable to contact you because of some issues. Ronak, would it be possible for you to stay in touch with me on whatsapp? I am unable to use my PC very often. I have an intention of starting a web dev business. I need some suggestions from you in this regard. But I need an easy and quick way of messaging. If you think it is okay for you, then please message me on my whatsapp +923461845071

**Comment by NooruddinLakhani** [comment]
You cannot randomly make a domain name; domains need to be purchased or obtained from a domain registrar. Websites like [register.domains](https://register.domains/en) can help you buy a custom domain if you don't have one already. They allow you to search for available domain names and purchase them.
